src/psyclone/psyir/tools/call_tree_utils.py

```diff
@@ -0,0 +1,415 @@
+        calls `get_input_parameter` and `get_output_parameter`, but avoids the
+        repeated computation of the variable usage.
+
+        :param node_list: list of PSyIR nodes to be analysed.
+        :type node_list: List[:py:class:`psyclone.psyir.nodes.Node`]
+        :param options: a dictionary with options for the CallTreeUtils
+            which will also be used when creating the VariablesAccessInfo
+            instance if required.
+        :type options: Optional[Dict[str, Any]]
+        :param Any options["COLLECT-ARRAY-SHAPE-READS"]: if this option is
+            set to a True value, arrays used as first parameter to the
+            PSyIR operators lbound, ubound, or size will be reported as
+            'read'. Otherwise, these accesses will be ignored.
+
+        :returns: a ReadWriteInfo object with the information about input-
+            and output parameters.
+        :rtype: :py:class:`psyclone.psyir.tools.ReadWriteInfo`
+
+        '''
+        variables_info = VariablesAccessInfo(node_list, options=options)
+        read_write_info = ReadWriteInfo()
+        self.get_input_parameters(read_write_info, node_list, variables_info)
+        self.get_output_parameters(read_write_info, node_list, variables_info)
+        return read_write_info
+
+    # -------------------------------------------------------------------------
+    def get_non_local_read_write_info(self, node_list, read_write_info):
+        '''Returns the information about non-local variables that are read
+        or written.
+
+        '''
+
+        # First collect all non-local symbols from the kernels called. They
+        # are collected in the todo list. This list will initially contain
+        # unknown accesses, since at this stage we cannot always differentiate
+        # between function calls and array accesses. In the resolve step
+        # that is following the corresponding modules will be queried and
+        # the right accesses (functions or variables) will be used.
+        todo = []
+        mod_manager = ModuleManager.get()
+        for node in node_list:
+            for kernel in node.walk(Kern):
+                if isinstance(kernel, BuiltIn):
+                    # Builtins don't have non-local accesses
+                    continue
+
+                # Get the non-local access information from the kernel
+                # by querying the module that contains the kernel:
+                try:
+                    mod_info = mod_manager.get_module_info(kernel.module_name)
+                except FileNotFoundError:
+                    print(f"Could not find module '{kernel.module_name}' - "
+                          f"ignored.")
+                    continue
+
+                all_possible_routines = mod_info.resolve_routine(kernel.name)
+                for routine_name in all_possible_routines:
+                    psyir = mod_info.get_psyir(routine_name)
```

Comment:
I think that `get_psyir` on a ModuleInfo object should return the psyir of the Module. That can then be queried for routines.

Reply:
I thought we had agreed that tools would act on the PSyIR, and not adding these functions to the PSyIR (to keep it small). But no problem, I've changed it by moving this function into Container.